### PR TITLE
Make file browser respond to focussed elements

### PR DESCRIPTION
**Pull request by gabalafou**
## References

- https://github.com/jupyterlab/jupyterlab/issues/9399
- #13415 - The main impetus for this PR

PR #13415 made focus visible, which as @krassowski [pointed out](https://github.com/jupyterlab/jupyterlab/pull/13415#issuecomment-1314551386):

> This surfaces an issue with the file browser ignoring focused elements.

In order to understand this PR, it's important to understand that "selection" and "focus" are not the same thing. The easiest way to understand the difference between the two is to play with the file browser in Windows or Ubuntu. If you have Windows or Ubuntu, and you open the file browser, you can move focus independent of selection by pressing `ctrl` + arrow key. Focus is visually indicated with a focus ring, while selection is indicated with a highlight, or background color. This PR basically adopts the focus-versus-selection patterns in Windows and Ubuntu for the JupyterLab file browser.

## Code changes

Most of the code changes in this PR are to the `DirListing` class in the `@jupyterlab/filebrowser` package. The main change is the addition of a new private state variable `_focusIndex`, which keeps track of where the focus should be in the directory listing.

Previously, every file or directory in the file browser had a tab index of 0. I changed it so that there should only ever be one file item node at a time that can be focussed with the tab key. This makes navigating the JupyterLab UI easier because the user won't have to tab through all of the items in the file browser to get to another part of the UI. 

I use the focus index in a few places to simplify `DirListing` methods. For example, the multi select method used to iterate through the directory listing to find the selected item nearest to the one clicked, but now it simply looks up the item at the focus index.

There are a number of other code changes to fix issues I uncovered with the file browser keyboard shortcuts. For example, I moved the `Backspace` keyboard shortcut out of the JSON schema file and into the `DirListing` class because it wasn't working when the `DirListing` was empty, and I couldn't think of a better way to fix it. 

### Test coverage

I added tests for about half of the code changes I made. Before adding more tests, I wanted to get this PR up for review, since changes requested in review could change the test code.

## User-facing changes

- Up/down arrow key, not tab key, moves focus. 
  - Moving keyboard focus from one file to another in the file browser is now exclusively done via the up and down arrow keys. Previously the user could move focus (but not selection) with the tab key (but it was useless because the file browser previously only responded to keyboard shortcuts on selected items, completely ignoring which item had focus; now it's not so easy to unintentionally move focus independently of selection). 
- `CTRL`+`Space` now selects/unselects files.
- `Backspace` now works to go up even when folder is empty.
- User can now do a keyboard-only discontiguous multi-select.
  - Previously, the only way to select discontinuous groups of files in the file browser was by shift-clicking with the mouse (or by enabling checkboxes in the settings editor and then clicking the checkboxes next to files). There was no way to do this with just the keyboard.
  - This is accomplished the same way that Ubuntu and Windows do it: `CTRL`+ up (or down) arrow key moves the focus. Then `CTRL`+`Space` toggles selection.
- Hopefully users will barely notice this, but the way multi-select is done is now different. For multi-select, I found inconsistencies between Mac, Windows, and Ubuntu, so I combined some characteristics that I found helpful in those platforms with some of my own choices. We will need people to try it out and see if it feels intuitive or not.
- When user renames a file, focus stays on that file after rename (or if the user presses the `esc` key).
- `Enter` keyboard shortcut can now open multiple files instead of just one.
- Commands that cannot operate on multiple files (such as "copy download link") do not appear in the context menu when user has selected multiple files.
- Focus stays in directory listing after deleting files.

## Backwards-incompatible changes

None, as far as I'm aware.

### CODE REVIEW

**Review by gabalafou — COMMENTED** [review]
Done self-reviewing!

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thank @gabalafou 

Here is a first quick review - I did not test it yet.

File: packages/filebrowser/src/listing.ts
Comment (gabalafou) [reply packages/filebrowser/src/listing.ts:2396]: Yep. Tested on Safari. Works the same.

File: packages/filebrowser/src/browser.ts
Comment (gabalafou) [reply packages/filebrowser/src/browser.ts]: Hehe. Agreed. I thought there was some reason for doing it the way I did but now I can't remember why, so I'll implement it the way you suggest.

File: packages/filebrowser/style/base.css
Comment (gabalafou) [reply packages/filebrowser/style/base.css]: Ok will do.

File: packages/filebrowser/style/base.css
Comment (gabalafou) [reply packages/filebrowser/style/base.css]: Ditto.

File: packages/filebrowser/test/listing.spec.ts
Comment (gabalafou) [reply packages/filebrowser/test/listing.spec.ts]: Oh good catch. That was my editor. Should be fixed now.

**Review by gabalafou — COMMENTED** [review]
Thanks for your review comments @fcollonval! I believe I addressed all of them. If I missed anything, it wasn't intentional, please bring it to my attention again.

I changed the CSS rule that controls the focus ring of the directory items to use theme variables and checked it in light-mode across a variety of browsers (Chrome, Safari, and Firefox on MacOS). I also checked dark-mode on MacOS Chrome. I probably need @isabela-pf or someone more familiar than me with the graphic design conventions of JupyterLab to give that a look and okay it before we merge this in.

**Comment by gabalafou** [comment]
Converting this PR to draft because I need to fix the last merge.

**Comment by gabalafou** [comment]
Okay, I fixed the merge conflicts. This should be ready for review again.

The labels for this PR might need to be fixed. When I was trying to fix the branch, I pushed a commit with a bunch of changes unrelated to this PR and I think it caused a bunch of labels to be added that shouldn't have been. I deleted many of them but I'm not sure if the remaining labels are correct.

File: packages/filebrowser-extension/schema/browser.json
Comment (gabalafou) [inline packages/filebrowser-extension/schema/browser.json:134]: Moved backspace handling into the DirListing class. You can reason from the way it's configured here that it doesn't work if the directory is empty, because if it's empty there is no `.jp-DirListing-itemText` node.

**Review by fcollonval — COMMENTED** [review]
Thanks @gabalafou 

I tested on Binder - it works great; I only uncover one corner case (but we may tackle it as a follow-up): if you open the context menu with the keyboard and then press _Escape_ to close it. The focus does not return on the filebrowser.

Could you fix the linter error and the JS filebrowser tests?

File: packages/filebrowser/src/opendialog.ts
Comment (gabalafou) [inline packages/filebrowser/src/opendialog.ts:151]: Out of curiosity, is `OpenDialog` still used? I can't seem to figure out how load any of this in the UI.

**Comment by gabalafou** [comment]
Thanks @fcollonval! I push two commits that should hopefully fix the test timeout and linter error.

My biggest concern before merging this PR is me [moving `Backspace` out of the settings-editable shortcuts and into the DirListing class](https://github.com/jupyterlab/jupyterlab/pull/13577/files#r1117512403). Could you comment on that specifically @fcollonval? Or perhaps @krassowski? since I believe @krassowski added the Backspace keyboard shortcut in #6859.

**Comment by fcollonval** [comment]
> My biggest concern before merging this PR is me [moving `Backspace` out of the settings-editable shortcuts and into the DirListing class](https://github.com/jupyterlab/jupyterlab/pull/13577/files#r1117512403). Could you comment on that specifically @fcollonval? Or perhaps @krassowski? since I believe @krassowski added the Backspace keyboard shortcut in #6859.

I don't have strong feeling about this. It seems that `Backspace` is the shortcut used by OS file browser. So I guess it is fine to get it hard coded. But let's wait for Mike feedback.

Regarding UI tests, they are currently broken on master.

**Comment by gabalafou** [comment]
@fcollonval I opened an issue to keep track of the bug you mentioned in your review and I will try to submit a fix in the near future:

- https://github.com/jupyterlab/jupyterlab/issues/14153

**Comment by krassowski** [comment]
I do not have strong feelings on hard-coding `Backspace` for "go-up" but it would be nice if we preserved this as dropping a shortcut will break user settings when upgrading and limit customisability (and we talked about accessibility overlapping with customisability on the call yesterday). I see this comment:

> You can reason from the way it's configured here that it doesn't work if the directory is empty, because if it's empty there is no `.jp-DirListing-itemText` node.

I believe we can just adjust the selector for the keybinding and it would just work. Is there any other reason for hard-coding it?

**Comment by gabalafou** [comment]
> I believe we can just adjust the selector for the keybinding and it would just work. Is there any other reason for hard-coding it?

No other reason. I just don't know how to do it. 😑

**Comment by gabalafou** [comment]
But I can give it another think and get back to you.

**Comment by isabela-pf** [comment]
I'm not finding any problems! I think this PR is well-scoped and I think everything happening with focus in the file browser makes sense to me. Just to note some positives:

- I'm happy to find the styling matches upcoming WCAG 2.2's more specific [focus appearance guidelines](https://www.w3.org/TR/WCAG22/#focus-appearance).
- I'm also happy to find the CSS changes seem mindful of the existing variables.
- It works in light and dark modes!

I am curious about the ways that the focus styling are different in the file browser versus other areas of the UI, but I do think that is outside the scope of the PR. Just noting it for follow up on my part.

A final note from me: `CTRL`+ up (or down) arrow key did not allow me to move focus separate from selection, but instead triggered something on my operating system that I'm not familiar with. If this isn't a problem for other Mac users, I won't push it on this PR because I know I'm likely not using the most recent MacOS.

**Comment by gabalafou** [comment]
Thanks @isabela-pf! I want to follow up on your note about `ctrl` + up/down. Can you describe what it triggered on your OS? Does it happen only in the browser or does it happen in any app? 

I would say that the keyboard-only, discontiguous multi-select feature is more aimed at Windows and Ubuntu users than Mac users, but all the same, I would like things to mostly work the same across operating systems, including older versions of Mac OS.

**Comment by gabalafou** [comment]
Also, to your note about focus styling, I would add that in my experience, JupyterLab's focus styling is not terribly consistent, and that this is an area that could use some polishing, but yes, I agree that it's a bigger scope than this PR. 

I wonder if a good first step towards consistency would be to add some variables for outline styles that could be reused across the app. I have the [variables.css](https://github.com/jupyterlab/jupyterlab/blob/master/packages/theme-light-extension/style/variables.css) file in mind. (That one links to the light theme version, but there is also a dark theme version of the file.) I imagine adding a section like:

```css
/* Outline (focus ring) */
--jp-outline-color: white;
--jp-outline-style: solid;
--jp-outline-width: 1px;
```

And for the dark theme version:

```css
/* Outline (focus ring) */
--jp-outline-color: #111;
--jp-outline-style: solid;
--jp-outline-width: 1px;
```

I'm not sure about those values, but... something like that to start.

File: packages/filebrowser-extension/schema/browser.json
Comment (gabalafou) [inline packages/filebrowser-extension/schema/browser.json:143]: Okay, this seems to do the trick: just add another selector to handle the case when the directory is empty and leave the other one alone. I wish I could leave a comment in this JSON file.

This `:focus` selector works for when the directory is empty because the only node (within the directory listing widget) that can receive focus at that point is the directory listing node itself. We cannot use the selector without `:focus` though because then it would get triggered if the user were in the middle of a file name, makes a typo, and then presses backspace to delete the typo. 

I tested this change three ways:

1. Opening a empty folder then pressing backspace
2. Opening a non-empty folder then pressing backspace
3. Renaming a file and pressing backspace while editing the file name

**Review by gabalafou — COMMENTED** [review]
@krassowski, I'm happy to report that I was able to keep the backspace as a user-configurable shortcut.

I think this PR is ready for merge now. 😄

**Review by fcollonval — APPROVED** [review]
Thanks the patience @gabalafou

**Review by fcollonval — APPROVED** [review]
Thanks @gabalafou

**Pull request merged**
